src/sidebar/helpers/mentions.ts

```diff
@@ -96,27 +96,53 @@ export function wrapDisplayNameMentions(
 // some assumptions about the tag content in order to be able to get away
 // with it. Specifically we assume that these tags will not contain `<` or
 // `>` in attribute values or the HTML content of the tag.
-const MENTION_TAG_RE = /<a[^>]\bdata-hyp-mention\b[^>]*>([^<]+)<\/a>/g;
+const MENTION_TAG_RE = /<a[^>]\bdata-hyp-mention\b[^>]*>@([^<]+)<\/a>/g;
+
+export type UnwrapMentionsOptions = {
+  text: string;
+  mentionMode: MentionMode;
+  mentions?: Mention[];
+};
 
 /**
  * Replace all mentions wrapped in the special `<a data-hyp-mention />` tag with
  * their plain-text representation.
+ *
  * The plain-text representation depends on the mention mode:
  * - `username`: @username
  * - `display-name`: @[Display Name]
+ *
+ * If a list of mentions is provided, the tag's userid will be matched against
+ * it, so that the plain-text version uses the most recent username or display
+ * name.
+ * If a list is not provided or the mention is not found, the tag's content will
+ * be used as username or display name.
  */
-export function unwrapMentions(text: string, mentionMode: MentionMode) {
+export function unwrapMentions({
+  text,
+  mentionMode,
+  mentions = [],
+}: UnwrapMentionsOptions) {
   // Use a regex rather than HTML parser to replace the mentions in order
   // to avoid modifying any of the content outside of the replaced tags. This
   // includes avoiding modifications such as encoding characters that will
   // happen when parsing and re-serializing HTML via eg. `innerHTML`.
-  return text.replace(MENTION_TAG_RE, (match, mention) => {
-    if (mentionMode === 'username') {
-      return mention;
-    }
+  return text.replace(MENTION_TAG_RE, (match, tagContent) => {
+    // Even though we need to capture the tag content via a regex for the
+    // reasons explained above, we can use regular DOM APIs to get the userid
+    const tempElement = document.createElement('div');
+    tempElement.innerHTML = match;
+    const userid = tempElement
+      .querySelector('a[data-hyp-mention]')
+      ?.getAttribute('data-userid');
+
+    const mention = mentions.find(
+      ({ original_userid }) => original_userid === userid,
+    );
 
-    const [atChar, ...rest] = mention;
-    return `${atChar}[${rest.join('')}]`;
+    return mentionMode === 'username'
+      ? `@${mention?.username ?? tagContent}`
+      : `@[${mention?.display_name || tagContent}]`;
```

Comment:
For the display name I used `||` instead of `??` so that, if the user used to have a display name when mention, and they have now removed it, we still show their original display name, otherwise the mention will become `@[]` and simply `@` when rendered.

This is an edge case that may need more treatment, perhaps not allowing users to remove their display name once set, but that's probably a problem to be treated on its own.

Reply:
This makes sense, but can you add this information in a brief comment.

Reply:
In the LMS context, users with empty display names are going to be a problem for mentions. We shouldn't have users with empty display names, since that would cause a bunch of other issues. It might be worth checking what happens if you try to set yourself to have an empty display name in various LMSes - I expect the backend code will generate one from the first name and last name instead?

Reply:
I reckon students with empty names should already be causing other types of problems, so I imagine this will not happen too frequently.